<a href="https://colab.research.google.com/github/SapientSapiens/ml-zoomcamp-2024/blob/main/homework/01/Homework01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
import numpy as np
import pandas as pd

#Q1. Pandas version

What's the version of Pandas that you installed?

In [3]:
pd.__version__

'2.1.4'

# Q2. Records count

How many records are in the dataset?

In [4]:
df = pd.read_csv('laptops.csv')

In [ ]:
df.shape

(2160, 12)

In [ ]:
df.shape[0]

2160

In [ ]:
len(df.index)

2160

Hence, there are 2160 records

# Q3. Laptop brands


How many laptop brands are presented in the dataset?

In [ ]:
df.head()

Laptop Status   Brand  \
0  ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...    New    Asus   
1  Alurin Go Start Intel Celeron N4020/8GB/256GB ...    New  Alurin   
2  ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...    New    Asus   
3  MSI Katana GF66 12UC-082XES Intel Core i7-1270...    New     MSI   
4  HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...    New      HP   

        Model            CPU  RAM  Storage Storage type       GPU  Screen  \
0  ExpertBook  Intel Core i5    8      512          SSD       NaN    15.6   
1          Go  Intel Celeron    8      256          SSD       NaN    15.6   
2  ExpertBook  Intel Core i3    8      256          SSD       NaN    15.6   
3      Katana  Intel Core i7   16     1000          SSD  RTX 3050    15.6   
4         15S  Intel Core i5   16      512          SSD       NaN    15.6   

  Touch  Final Price  
0    No      1009.00  
1    No       299.00  
2    No       789.00  
3    No      1199.00  
4    No       669.01

In [ ]:
df['Brand'].nunique()

27

In [ ]:
# the distinct laptop brands are

df['Brand'].unique()

array(['Asus', 'Alurin', 'MSI', 'HP', 'Lenovo', 'Medion', 'Acer', 'Apple',
       'Razer', 'Gigabyte', 'Dell', 'LG', 'Samsung', 'PcCom', 'Microsoft',
       'Primux', 'Prixton', 'Dynabook Toshiba', 'Thomson', 'Denver',
       'Deep Gaming', 'Vant', 'Innjoo', 'Jetwing', 'Millenium', 'Realme',
       'Toshiba'], dtype=object)

# Q4. Missing values

How many columns in the dataset have missing values?

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2160 entries, 0 to 2159
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Laptop        2160 non-null   object 
 1   Status        2160 non-null   object 
 2   Brand         2160 non-null   object 
 3   Model         2160 non-null   object 
 4   CPU           2160 non-null   object 
 5   RAM           2160 non-null   int64  
 6   Storage       2160 non-null   int64  
 7   Storage type  2118 non-null   object 
 8   GPU           789 non-null    object 
 9   Screen        2156 non-null   float64
 10  Touch         2160 non-null   object 
 11  Final Price   2160 non-null   float64
dtypes: float64(2), int64(2), object(8)
memory usage: 202.6+ KB


## From the above, it is evident that 3 columns viz. Storage Type, GPU and Screen have null values


## Programatically, we can find out as below

In [ ]:
i =0
for col in df.columns:
 if df[col].isnull().sum() > 0:
  i+=1
  print(i,'.',col)

1 . Storage type
2 . GPU
3 . Screen


# Q5. Maximum final price

What's the maximum final price of Dell notebooks in the dataset?

In [ ]:
df[df['Brand'] == 'Dell']['Final Price'].max()

3936.0

In [ ]:
# alternatively we can write

df.query('Brand == "Dell"')['Final Price'].max()

3936.0

In [ ]:
# or we can also write

df.groupby('Brand')['Final Price'].max()['Dell']

3936.0

# Q6. Median value of Screen


1.Find the median value of Screen column in the dataset.

2.Next, calculate the most frequent value of the same Screen column.

3.Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.

4.Now, calculate the median value of Screen once again.


# Has it changed?

 ### Hint: refer to existing mode and median functions to complete the task.

   ### Yes
   ### No

## 1.Find the median value of Screen column in the dataset.

In [ ]:
df['Screen'].sort_values(ascending=True).median()

15.6

## 2.Next, calculate the most frequent value of the same Screen column.

In [ ]:
# the most frequest or occuring value will be the mode in that value range

df['Screen'].mode()

0    15.6
Name: Screen, dtype: float64

In [ ]:
df[df['Screen'] == 15.6].shape[0]

1009

##### Almost 50% of the values are 15.6 in this column

## 3.Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.

In [5]:
# before the fillna execution
df[df['Screen'].isnull()]

Laptop Status   Brand  \
624   Acer Extensa 15 EX215-54 Intel Core i5-1135G7/...    New    Acer   
1430  HP ENVY x360 2-in-1 Laptop 15-ew0008np Intel C...    New      HP   
1503  Lenovo IdeaPad Gaming 3 15ACH6 AMD Ryzen 5 560...    New  Lenovo   
1548  Lenovo ThinkPad P15 Gen 2 Intel Core i7-11850H...    New  Lenovo   

         Model            CPU  RAM  Storage Storage type        GPU  Screen  \
624    Extensa  Intel Core i5    8      256          SSD        NaN     NaN   
1430      Envy  Intel Core i7   16      512          SSD   RTX 2050     NaN   
1503   IdeaPad    AMD Ryzen 5   16      512          SSD   RTX 3060     NaN   
1548  ThinkPad  Intel Core i7   16      512          SSD  RTX A2000     NaN   

     Touch  Final Price  
624     No       524.99  
1430   Yes      1863.52  
1503    No      1505.00  
1548    No      2569.00

In [6]:
# filling missing values with fillna method with 15.6 got from previous step
df['Screen'].fillna(15.6, inplace=True)

In [7]:
# after the fillna execution
df[df['Screen'].isnull()]

Empty DataFrame
Columns: [Laptop, Status, Brand, Model, CPU, RAM, Storage, Storage type, GPU, Screen, Touch, Final Price]
Index: []

# 4.Now, calculate the median value of Screen once again.

In [ ]:
df['Screen'].sort_values(ascending=True).median()

15.6

# So, no, the meidan has not changed

# Q7. Sum of weights


  1.Select all the "Innjoo" laptops from the dataset.

  2.Select only columns RAM, Storage, Screen.

  3.Get the underlying NumPy array. Let's call it X.

  4.Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

  5.Compute the inverse of XTX.

  6.Create an array y with values [1100, 1300, 800, 900, 1000, 1100].

  7.Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

  8.What's the sum of all the elements of the result?




Note: You just implemented linear regression. We'll talk about it in the next lesson.

 0.43

 45.29

 45.58

 91.30

# 1.Select all the "Innjoo" laptops from the dataset.

In [ ]:
df[df['Brand'] == 'Innjoo']

Laptop Status   Brand Model  \
1478  InnJoo Voom Excellence Intel Celeron N4020/8GB...    New  Innjoo  Voom   
1479  InnJoo Voom Excellence Pro Intel Celeron N4020...    New  Innjoo  Voom   
1480  Innjoo Voom Intel Celeron N3350/4GB/64GB eMMC/...    New  Innjoo  Voom   
1481  Innjoo Voom Laptop Max Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1482  Innjoo Voom Laptop Pro Intel Celeron N3350/6GB...    New  Innjoo  Voom   
1483  Innjoo Voom Pro Intel Celeron N3350/6GB/128GB ...    New  Innjoo  Voom   

                CPU  RAM  Storage Storage type  GPU  Screen Touch  Final Price  
1478  Intel Celeron    8      256          SSD  NaN    15.6    No       311.37  
1479  Intel Celeron    8      512          SSD  NaN    15.6    No       392.55  
1480  Intel Celeron    4       64         eMMC  NaN    14.1    No       251.40  
1481  Intel Celeron    6       64         eMMC  NaN    14.1    No       383.61  
1482  Intel Celeron    6      128          SSD  NaN    14.1    No       317.02  
1483  Intel Celeron    6      128         eMMC  NaN    14.1    No       431.38

# 2.Select only columns RAM, Storage, Screen.

In [ ]:
df[df['Brand'] == 'Innjoo'][['RAM', 'Storage', 'Screen']]

RAM  Storage  Screen
1478    8      256    15.6
1479    8      512    15.6
1480    4       64    14.1
1481    6       64    14.1
1482    6      128    14.1
1483    6      128    14.1

# 3.Get the underlying NumPy array. Let's call it X.

In [8]:
X = df[df['Brand'] == 'Innjoo'][['RAM', 'Storage', 'Screen']].to_numpy()

In [9]:
X

array([[  8. , 256. ,  15.6],
       [  8. , 512. ,  15.6],
       [  4. ,  64. ,  14.1],
       [  6. ,  64. ,  14.1],
       [  6. , 128. ,  14.1],
       [  6. , 128. ,  14.1]])

# 4.Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.

In [47]:
# the transposed matrix of T
XT = X.T

In [56]:
# matrix multiplication product of X.T and X
XTX = np.matmul(XT, X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [57]:
# the notation @ can also be used for np.matmul function.
XTX = XT@X
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

In [58]:
# checking with the product function also yeilds same result
XTX = XT.dot(X)
XTX

array([[2.52000e+02, 8.32000e+03, 5.59800e+02],
       [8.32000e+03, 3.68640e+05, 1.73952e+04],
       [5.59800e+02, 1.73952e+04, 1.28196e+03]])

# 5.Compute the inverse of XTX.

In [63]:
# inverse of XTX
XTX_inv = np.linalg.inv(XTX)

In [64]:
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

# 6.Create an array y with values [1100, 1300, 800, 900, 1000, 1100].

In [61]:
list = [1100, 1300, 800, 900, 1000, 1100]
y = np.array(list)

In [62]:
y

array([1100, 1300,  800,  900, 1000, 1100])

# 7.Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.

In [65]:
XTX_inv

array([[ 2.78025381e-01, -1.51791334e-03, -1.00809855e-01],
       [-1.51791334e-03,  1.58286725e-05,  4.48052175e-04],
       [-1.00809855e-01,  4.48052175e-04,  3.87214888e-02]])

In [66]:
XT

array([[  8. ,   8. ,   4. ,   6. ,   6. ,   6. ],
       [256. , 512. ,  64. ,  64. , 128. , 128. ],
       [ 15.6,  15.6,  14.1,  14.1,  14.1,  14.1]])

In [67]:
# Multiplying the inverse of XTX with the transpose of X
w_halfway = np.matmul(XTX_inv, XT)

In [70]:
w_halfway

array([[ 0.26298349, -0.12560233, -0.40646389,  0.14958687,  0.05244042,
         0.05244042],
       [-0.00110155,  0.00295059,  0.00125892, -0.00177691, -0.00076387,
        -0.00076387],
       [-0.08772226,  0.0269791 ,  0.17140891, -0.0302108 , -0.00153546,
        -0.00153546]])

In [76]:
y

array([1100, 1300,  800,  900, 1000, 1100])

In [80]:
# Multplying the result by y
w = np.dot(w_halfway, y)

In [81]:
w

array([45.58076606,  0.42783519, 45.29127938])

# 8.What's the sum of all the elements of the result?

In [83]:
w.sum().round(2)

91.3